File: builder/src/services/normalizer.ts
import { CanonicalPortfolio, Project, Skill, Experience, Education, Achievement } from "../types/portfolio";
import { generateId, sanitizeUrl } from "../lib/utils";

export const DEFAULT_TEMPLATE = "dark_developer";

export function createEmptyPortfolio(template = DEFAULT_TEMPLATE): CanonicalPortfolio {
  return {
    selected_template: template,
    full_name: "",
    headline: "",
    bio: "",
    profile_image_url: "",
    location: "",
    email: "",
    github_url: "",
    linkedin_url: "",
    resume_url: "",
    resume_file_name: "",
    projects: [],
    skills: [],
    experiences: [],
    education: [],
    achievements: [],
  };
}

export function normalizePortfolio(raw: any, fallbackTemplate = DEFAULT_TEMPLATE): CanonicalPortfolio {
  if (!raw || typeof raw !== "object") {
    return createEmptyPortfolio(fallbackTemplate);
  }

  const selected_template = typeof raw.selected_template === "string" && raw.selected_template.trim()
    ? raw.selected_template.trim()
    : fallbackTemplate;

  const full_name = typeof raw.full_name === "string" ? raw.full_name.trim() : (typeof raw.name === "string" ? raw.name.trim() : "");
  const headline = typeof raw.headline === "string" ? raw.headline.trim() : (typeof raw.role === "string" ? raw.role.trim() : "");
  const bio = typeof raw.bio === "string" ? raw.bio.trim() : (typeof raw.about === "string" ? raw.about.trim() : (typeof raw.summary === "string" ? raw.summary.trim() : ""));
  
  const profile_image_url = typeof raw.profile_image_url === "string" ? raw.profile_image_url.trim() : (typeof raw.avatar_url === "string" ? raw.avatar_url.trim() : "");
  const location = typeof raw.location === "string" ? raw.location.trim() : (typeof raw.city === "string" ? raw.city.trim() : "");
  const email = typeof raw.email === "string" ? raw.email.trim() : (typeof raw.contact_email === "string" ? raw.contact_email.trim() : "");

  const github_url = sanitizeUrl(typeof raw.github_url === "string" ? raw.github_url : (raw.social?.github || raw.github || ""));
  const linkedin_url = sanitizeUrl(typeof raw.linkedin_url === "string" ? raw.linkedin_url : (raw.social?.linkedin || raw.linkedin || ""));
  const resume_url = sanitizeUrl(typeof raw.resume_url === "string" ? raw.resume_url : "");
  const resume_file_name = typeof raw.resume_file_name === "string" ? raw.resume_file_name.trim() : "";

  // Normalize Projects
  const projects: Project[] = Array.isArray(raw.projects)
    ? raw.projects.map((p: any, idx: number) => {
        const title = typeof p?.title === "string" ? p.title.trim() : (typeof p?.name === "string" ? p.name.trim() : `Project ${idx + 1}`);
        const description = typeof p?.description === "string" ? p.description.trim() : (typeof p?.desc === "string" ? p.desc.trim() : "");
        const image_url = typeof p?.image_url === "string" ? p.image_url.trim() : (typeof p?.src === "string" ? p.src.trim() : (typeof p?.img === "string" ? p.img.trim() : ""));
        const technologies = Array.isArray(p?.technologies)
          ? p.technologies.map((t: any) => String(t).trim()).filter(Boolean)
          : (Array.isArray(p?.tags) ? p.tags.map((t: any) => String(t?.name || t).trim()).filter(Boolean) : []);
        const github_url = sanitizeUrl(typeof p?.github_url === "string" ? p.github_url : (p?.github || p?.sourceCodeLink || ""));
        const live_url = sanitizeUrl(typeof p?.live_url === "string" ? p.live_url : (p?.live || p?.liveLink || p?.link || ""));
        
        return {
          id: p?.id ? String(p.id) : generateId(`project-${idx + 1}`),
          title,
          description,
          image_url,
          technologies,
          github_url,
          live_url,
        };
      })
    : [];

  // Normalize Skills
  const skills: Skill[] = Array.isArray(raw.skills)
    ? raw.skills.map((s: any, idx: number) => {
        if (typeof s === "string") {
          return {
            id: generateId(`skill-${idx + 1}`),
            name: s.trim(),
            level: "Advanced",
          };
        }
        return {
          id: s?.id ? String(s.id) : generateId(`skill-${idx + 1}`),
          name: typeof s?.name === "string" ? s.name.trim() : (typeof s?.title === "string" ? s.title.trim() : `Skill ${idx + 1}`),
          level: s?.level || "Advanced",
          category: s?.category || undefined,
        };
      })
    : [];

  // Normalize Experiences
  const experiences: Experience[] = Array.isArray(raw.experiences)
    ? raw.experiences.map((e: any, idx: number) => {
        const company = typeof e?.company === "string" ? e.company.trim() : (typeof e?.companyName === "string" ? e.companyName.trim() : `Company ${idx + 1}`);
        const role = typeof e?.role === "string" ? e.role.trim() : (typeof e?.title === "string" ? e.title.trim() : "Software Engineer");
        const start_date = typeof e?.start_date === "string" ? e.start_date.trim() : (typeof e?.startDate === "string" ? e.startDate.trim() : (typeof e?.date === "string" ? e.date.split("—")[0]?.trim() || e.date : ""));
        const end_date = typeof e?.end_date === "string" ? e.end_date.trim() : (typeof e?.endDate === "string" ? e.endDate.trim() : (typeof e?.date === "string" && e.date.includes("—") ? e.date.split("—")[1]?.trim() || "Present" : "Present"));
        const description = typeof e?.description === "string" 
          ? e.description.trim() 
          : (Array.isArray(e?.points) ? e.points.join("\n") : (Array.isArray(e?.description) ? e.description.join("\n") : ""));

        return {
          id: e?.id ? String(e.id) : generateId(`exp-${idx + 1}`),
          company,
          role,
          start_date,
          end_date,
          description,
        };
      })
    : [];

  // Normalize Education
  const education: Education[] = Array.isArray(raw.education)
    ? raw.education.map((ed: any, idx: number) => {
        return {
          id: ed?.id ? String(ed.id) : generateId(`edu-${idx + 1}`),
          institution: typeof ed?.institution === "string" ? ed.institution.trim() : (typeof ed?.school === "string" ? ed.school.trim() : (typeof ed?.university === "string" ? ed.university.trim() : "")),
          degree: typeof ed?.degree === "string" ? ed.degree.trim() : "",
          field: typeof ed?.field === "string" ? ed.field.trim() : (typeof ed?.major === "string" ? ed.major.trim() : ""),
          start_year: typeof ed?.start_year === "string" ? ed.start_year.trim() : (typeof ed?.startYear === "string" ? ed.startYear.trim() : ""),
          end_year: typeof ed?.end_year === "string" ? ed.end_year.trim() : (typeof ed?.endYear === "string" ? ed.endYear.trim() : ""),
          description: typeof ed?.description === "string" ? ed.description.trim() : (typeof ed?.details === "string" ? ed.details.trim() : ""),
        };
      })
    : [];

  // Normalize Achievements
  const achievements: Achievement[] = Array.isArray(raw.achievements)
    ? raw.achievements.map((a: any, idx: number) => {
        return {
          id: a?.id ? String(a.id) : generateId(`ach-${idx + 1}`),
          title: typeof a?.title === "string" ? a.title.trim() : "",
          date: typeof a?.date === "string" ? a.date.trim() : (typeof a?.year === "string" ? a.year.trim() : ""),
          issuer: typeof a?.issuer === "string" ? a.issuer.trim() : (typeof a?.organization === "string" ? a.organization.trim() : ""),
          description: typeof a?.description === "string" ? a.description.trim() : "",
          credential_url: sanitizeUrl(typeof a?.credential_url === "string" ? a.credential_url : (a?.link || "")),
        };
      })
    : [];

  return {
    selected_template,
    full_name,
    headline,
    bio,
    profile_image_url,
    location,
    email,
    github_url,
    linkedin_url,
    resume_url,
    resume_file_name,
    projects,
    skills,
    experiences,
    education,
    achievements,
  };
}

File: OCR model api/normalizer.py
"""
Resume Normalizer for StackFolio Resume Intelligence.
Cleans whitespace, deduplicates skills, normalizes date formats,
and strictly sanitizes URLs according to the protocol security whitelist.
"""

import re
from typing import Dict, Any, List
from parser_deterministic import RawParsedResume

# Whitelisted protocols for external URLs
ALLOWED_URL_PROTOCOLS = ("http://", "https://")
PROHIBITED_URL_SCHEMES = ("javascript:", "data:", "vbscript:", "file:", "about:")

# Canonical Skill Mappings for safe deduplication
SKILL_CANONICAL_MAP = {
    "react": "React",
    "react.js": "React",
    "reactjs": "React",
    "react js": "React",
    "node": "Node.js",
    "node.js": "Node.js",
    "nodejs": "Node.js",
    "node js": "Node.js",
    "express": "Express.js",
    "express.js": "Express.js",
    "expressjs": "Express.js",
    "postgres": "PostgreSQL",
    "postgresql": "PostgreSQL",
    "mongo": "MongoDB",
    "mongodb": "MongoDB",
    "vue": "Vue.js",
    "vue.js": "Vue.js",
    "vuejs": "Vue.js",
    "tailwind": "Tailwind CSS",
    "tailwindcss": "Tailwind CSS",
    "tailwind css": "Tailwind CSS",
    "ts": "TypeScript",
    "typescript": "TypeScript",
    "js": "JavaScript",
    "javascript": "JavaScript",
    "python": "Python",
    "golang": "Go",
    "go": "Go",
    "docker": "Docker",
    "k8s": "Kubernetes",
    "kubernetes": "Kubernetes",
    "aws": "AWS",
    "git": "Git",
    "github": "GitHub",
    "next": "Next.js",
    "next.js": "Next.js",
    "nextjs": "Next.js",
}


def sanitize_url(raw_url: str) -> str:
    """
    Sanitizes extracted URLs.
    Allows only http:// and https://.
    Rejects javascript:, data:, vbscript:.
    Auto-prepends https:// to clean domains.
    """
    if not raw_url or not isinstance(raw_url, str):
        return ""

    trimmed = raw_url.strip()
    lower_trimmed = trimmed.lower()

    # Reject disallowed protocols
    for scheme in PROHIBITED_URL_SCHEMES:
        if lower_trimmed.startswith(scheme):
            return ""

    if lower_trimmed.startswith("http://") or lower_trimmed.startswith("https://"):
        return trimmed

    # Auto-prepend https:// to valid domain patterns
    if re.match(r"^[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}(?:/[^\s]*)?$", trimmed):
        return f"https://{trimmed}"

    return ""


def clean_text(text: Optional[str] if "Optional" in globals() else Any) -> str:
    """
    Normalizes multi-line whitespace into clean single spaces or paragraphs.
    """
    if not text or not isinstance(text, str):
        return ""
    # Consolidate multiple spaces into one
    return re.sub(r"\s+", " ", text).strip()


def normalize_skill_name(raw_name: str) -> str:
    """
    Normalizes skill name according to canonical dictionary while avoiding improper merges.
    """
    clean_n = clean_text(raw_name)
    key = clean_n.lower()
    return SKILL_CANONICAL_MAP.get(key, clean_n)


def normalize_parsed_resume(raw: RawParsedResume) -> Dict[str, Any]:
    """
    Normalizes all extracted sections into a structured schema dictionary.
    """
    # 1. Normalize Profile
    profile = {
        "full_name": clean_text(raw.full_name),
        "headline": clean_text(raw.headline),
        "bio": clean_text(raw.bio),
        "location": clean_text(raw.location),
        "email": raw.email.strip().lower() if raw.email else "",
        "github_url": sanitize_url(raw.github_url),
        "linkedin_url": sanitize_url(raw.linkedin_url),
    }

    # 2. Normalize Experiences
    normalized_experiences: List[Dict[str, Any]] = []
    for idx, exp in enumerate(raw.experiences):
        company = clean_text(exp.get("company", ""))
        role = clean_text(exp.get("role", ""))
        if company or role:
            normalized_experiences.append({
                "company": company or "Company",
                "role": role or "Software Engineer",
                "start_date": clean_text(exp.get("start_date", "")),
                "end_date": clean_text(exp.get("end_date", "")),
                "description": clean_text(exp.get("description", "")),
                "display_order": idx
            })

    # 3. Normalize Education
    normalized_education: List[Dict[str, Any]] = []
    for idx, edu in enumerate(raw.education):
        institution = clean_text(edu.get("institution", ""))
        degree = clean_text(edu.get("degree", ""))
        if institution or degree:
            normalized_education.append({
                "institution": institution or "University",
                "degree": degree or "Degree",
                "field": clean_text(edu.get("field", "")),
                "start_year": clean_text(edu.get("start_year", "")),
                "end_year": clean_text(edu.get("end_year", "")),
                "description": clean_text(edu.get("description", "")),
                "display_order": idx
            })

    # 4. Normalize Projects
    normalized_projects: List[Dict[str, Any]] = []
    for idx, proj in enumerate(raw.projects):
        title = clean_text(proj.get("title", ""))
        if title:
            techs = proj.get("technologies", [])
            clean_techs = [clean_text(t) for t in techs if clean_text(t)]
            normalized_projects.append({
                "title": title,
                "description": clean_text(proj.get("description", "")),
                "technologies": clean_techs,
                "github_url": sanitize_url(proj.get("github_url", "")),
                "live_url": sanitize_url(proj.get("live_url", "")),
                "display_order": idx
            })

    # 5. Normalize Skills (deduplicate by canonical name)
    seen_skills = set()
    normalized_skills: List[Dict[str, Any]] = []
    for exp_skill in raw.skills:
        raw_name = exp_skill.get("name", "")
        norm_name = normalize_skill_name(raw_name)
        if norm_name and norm_name.lower() not in seen_skills:
            seen_skills.add(norm_name.lower())
            normalized_skills.append({
                "name": norm_name,
                "category": clean_text(exp_skill.get("category", "Technical")),
                "level": clean_text(exp_skill.get("level", "Intermediate")),
                "display_order": len(normalized_skills)
            })

    # 6. Normalize Achievements
    normalized_achievements: List[Dict[str, Any]] = []
    for idx, ach in enumerate(raw.achievements):
        title = clean_text(ach.get("title", ""))
        if title:
            normalized_achievements.append({
                "title": title,
                "issuer": clean_text(ach.get("issuer", "")),
                "date": clean_text(ach.get("date", "")),
                "description": clean_text(ach.get("description", "")),
                "credential_url": sanitize_url(ach.get("credential_url", "")),
                "display_order": idx
            })

    return {
        "profile": profile,
        "experiences": normalized_experiences,
        "education": normalized_education,
        "projects": normalized_projects,
        "skills": normalized_skills,
        "achievements": normalized_achievements
    }
